### PR TITLE
HEEDLS-345 Fix incorrect boolean condition

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
Fixed an incorrect condition causing email validation to fail when unchanged. Added a couple of unit tests to check it.

### CODE REVIEW

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
This all looks good to me, but there is one thing that will need changing.

I double checked, and you can indeed change a delegate account email address to the email address of another delegate account at the same centre, but the condition in the validation method should clearly exclude that. But I've figured out why it gets bypassed - the GetDelegateUsersByEmailAddress method (called as part of the GetUsersByEmailAddress) does not return the CentreId, so the delegateUsers populated by it will have a default CentreId (probably 0). Thus the non-zero centre Id passed in will never be the same, and thus the email address change occurs.

So the GetDelegateUsersByEmailAddress in the UserDataService will need a quick change to make sure the centre id is recovered for the users so it can be used in the validation.

**Comment by AlexJacksonDS** [comment]
Got to love queries that don't fully populate objects. Will get that changed.

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 1]
```diff
@@ -144,7 +144,7 @@ public List<CentreUserDetails> GetUserCentres(AdminUser? adminUser, List<Delegat
         public bool NewEmailAddressIsValid(string emailAddress, int? adminUserId, int? delegateUserId, int centreId)
         {
             var (adminUser, delegateUser) = GetUsersById(adminUserId, delegateUserId);
-            if (adminUser?.EmailAddress == emailAddress && delegateUser?.EmailAddress == emailAddress)
+            if ((adminUser == null || adminUser.EmailAddress == emailAddress) && (delegateUser == null || delegateUser.EmailAddress == emailAddress))
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/UserService.cs:147]: This statement is getting quite long - maybe worth pulling out into a method that's called `userEmailHasChanged(user, email)` or similar and doing if `(!userEmailHasChanged(adminUser, emailAddress) && !userEmailHasChanged(delegateUser, emailAddress) `

**Review by DanBloxham-sw — APPROVED** [review]
Requested change made.

File: DigitalLearningSolutions.Data/DataServices/UserDataService.cs [hunk 2]
```diff
@@ -181,14 +181,20 @@ public List<DelegateUser> GetDelegateUsersByEmailAddress(string emailAddress)
...
+                        cd.CandidateNumber,
+                        cd.CentreID,
+                        ct.CentreName,
+                        ct.Active AS CentreActive,
+                        cd.EmailAddress,
+                        cd.FirstName,
+                        cd.LastName,
+                        cd.Password,
+                        cd.Approved,
+                        cd.ResetPasswordID
```
Comment (stellake) [inline DigitalLearningSolutions.Data/DataServices/UserDataService.cs:194]: Do we need to change any test to check that these all come through fine now?

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/UserDataService.cs:194]: There are actually no tests for the GetByEmail methods. Should I write some as part of this change?

**Review by stellake — APPROVED** [review]
Looks good! 👍

**Pull request merged**